massa-bootstrap/src/server.rs

```diff
@@ -269,29 +254,18 @@ impl BootstrapServer {
                         let network_command_sender = self.network_command_sender.clone();
                         let config = self.bootstrap_config.clone();
 
-                        bootstrap_sessions.push(async move {
-                            debug!("awaiting on bootstrap of peer {}", remote_addr);
-                            match tokio::time::timeout(config.bootstrap_timeout.into(), manage_bootstrap(&config, &mut server, data_execution, version, consensus_command_sender, network_command_sender)).await {
-                                Ok(mgmt) => match mgmt {
-                                    Ok(_) => {
-                                        info!("bootstrapped peer {}", remote_addr)
-                                    },
-                                    Err(BootstrapError::ReceivedError(error)) => debug!("bootstrap serving error received from peer {}: {}", remote_addr, error),
-                                    Err(err) => {
-                                        debug!("bootstrap serving error for peer {}: {}", remote_addr, err);
-                                        // We allow unused result because we don't care if an error is thrown when sending the error message to the server we will close the socket anyway.
-                                        let _ = tokio::time::timeout(config.write_error_timeout.into(), server.send(BootstrapServerMessage::BootstrapError { error: err.to_string() })).await;
-                                    },
-                                }
-                                Err(_timeout) => {
-                                    debug!("bootstrap timeout for peer {}", remote_addr);
-                                    // We allow unused result because we don't care if an error is thrown when sending the error message to the server we will close the socket anyway.
-                                    let _ = tokio::time::timeout(config.write_error_timeout.into(), server.send(BootstrapServerMessage::BootstrapError { error: format!("Bootstrap process timedout ({})", format_duration(config.bootstrap_timeout.to_duration())) })).await;
-                                }
-                            }
-
-                        });
-                        massa_trace!("bootstrap.session.started", {"active_count": bootstrap_sessions.len()});
+                        tokio::spawn(run_bootstrap_session(
```

Comment:
If we stop the node this task will block the join of the main thread ?

Reply:
I've brought in a lot of follow-up work done in benph/bootstrap-de-async branch.

The approach taken is that this is now a `std::thread::spawn`, with `run_bootstrap_session` now a non-`async` function.

`run_bootstrap_session`, however, has a lot of async stuff that still needs to be removed. In the meantime, the compromise is to follow this pattern:

```rust
fn run_bootstrap_session(...) -> ... {
    tokio::Runtime::new().block_on(/* do the actual session here */);
}
```

It's not pretty, and if there's a better way to approach the idea in my mind (explained below), I'm all ears.

The goal is as follows:

- the `launch()` function is the top-level runtime context
- the `start_bootstrap_server` function runs in that context, but that is (close to) being non-`async`, and can be thought of as a such
- this function, in turn, spawns a thread running the servers main-loop
- this main-loop spawns a thread for each bootstrap session to be run
 - _inside_ this session, a tokio runtime is created so the remaining async code base can work
- As far as the the execution stack/contexts outside of this thread are concerned, there exists a thread for each bootstrap session, and it has the usuall block/unblock behaviour. They are oblivious to the fact that this is often done by that threads `tokio` `Runtime` context.

Reply:
It's very clear like that ! Thank you